src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala

```diff
@@ -186,63 +198,87 @@ class AppsResource @Inject() (
     }
   }
 
+  @SuppressWarnings(Array("all")) /* async/await */
   @PUT
   @Path("""{id:.+}""")
   def replace(
     @PathParam("id") id: String,
     body: Array[Byte],
     @DefaultValue("false")@QueryParam("force") force: Boolean,
     @DefaultValue("true")@QueryParam("partialUpdate") partialUpdate: Boolean,
-    @Context req: HttpServletRequest): Response = authenticated(req) { implicit identity =>
+    @Context req: HttpServletRequest,
+    @Suspended asyncResponse: AsyncResponse): Unit = sendResponse(asyncResponse) {
+    async {
+      implicit val identity = await(authenticatedAsync(req))
```

Comment:
Hm, why not

```scala
@Suspended asyncResponse: AsyncResponse): Unit = authenticatedAsync(req, asyncResponse) { implicit identity =>
   update(id, body, force, partialUpdate, req, allowCreation = true
}
```

So we let `authenticatedAsync` handle the async response. This reduces the number of changes again.

Reply:
Disagreed.

Reply:
We can do

```
authenticatedAsync(req, asyncResponse).map { implicit identity =>
}
```

I am opposed to passing a callback to authenticatedAsync, and having authenticatedAsync return a response. Reasons:

- It encourages nested callback hell.
- The callbacks are not monadic (you can not combine them with map/flatMap as you can a Future)
- Async / await does not work with futures inside callbacks. This means we have to have a new declaration of async for each callback
- Having authenticatedAsync return a response creates two pathways that we need to render auth rejection responses. By only using rejections, we collapse these to a single way.

Reply:
Good points. It's closer to what we have, though. Let's go with your solution then 🙂.

Since the diff is so big it'll take some time to review it :/.